File: day-14/part-2/ayc0.js
const { performance } = require("perf_hooks");

let mask = "XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX";
const memory = new Map();
const allOnes = BigInt(2 ** 36 - 1);

/**
 *
 * @param {string} s
 * @returns
 */
const run = (s) => {
  for (const line of s.trim().split("\n")) {
    // mask = "<>"
    if (line[1] === "a") {
      mask = line.substring(7, line.length);
      //   console.log(mask);
      continue;
    }

    // mem[<entry>] = <value>
    var indexOfEndBracket = line.indexOf("]");
    if (indexOfEndBracket == -1) continue;
    // to catch the values in "mem[<entry>] = <value>
    const entry = BigInt(line.substring(4, indexOfEndBracket));
    const value = Number(line.substring(indexOfEndBracket + 4, line.length));
    // console.log(entry, "<-", value);

    let correctedEntry = entry;
    const floatings = [];

    // 36 is the length of the mask
    let thisMask;
    for (let i = 0; i < 36; i++) {
      thisMask = mask[35 - i];
      if (thisMask === "0") {
        continue;
      } else if (thisMask === "1") {
        correctedEntry = correctedEntry | (1n << BigInt(i));
      } else {
        floatings.push(BigInt(i));
      }
    }

    // console.log(floatings);

    let correctedEntries = [correctedEntry];
    let tmpCorrectedEntries = [];
    for (const floating of floatings) {
      for (const correctedEntry of correctedEntries) {
        // marked bit at floating to 0
        tmpCorrectedEntries.push(correctedEntry & (allOnes - (1n << floating)));
        // marked bit at floating to 1
        tmpCorrectedEntries.push(correctedEntry | (1n << floating));
      }
      correctedEntries = tmpCorrectedEntries;
      tmpCorrectedEntries = [];
    }

    // console.log(correctedEntries);
    for (const correctedEntry of correctedEntries) {
      memory.set(correctedEntry, value);
    }
  }

  let total = 0;
  for (const value of memory.values()) {
    total += value;
  }
  return total;
};

let start = performance.now();
let input = process.argv[2];
// input = `
// mask = 000000000000000000000000000000X1001X
// mem[42] = 100
// mask = 00000000000000000000000000000000X0XX
// mem[26] = 1
// `;

let answer = run(input);

console.log(`_duration:${performance.now() - start}`);
console.log(answer);

File: day-14/part-2/ayc0.nim
from times import cpuTime
from os import paramStr
import strutils
import bitops
import tables

var mask = "XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX"
var memory = newTable[uint64, uint64]()
var allOnes = 0b1111_1111_1111_1111_1111_1111_1111_1111_1111_1111_1111_1111_1111_1111_1111_1111'u64

proc run(s: string): uint64 =
    var lines = splitLines(strip(s))

    for line in lines:
        if line[1] == 'a':
            # mask = "<>"
            mask = line[7 .. line.len - 1]
            # echo mask
            continue

        # mem[<entry>] = <value>
        var indexOfEndBracket = line.find(']')
        if indexOfEndBracket == -1:
            continue
        # to catch the values in "mem[<entry>] = <value>
        var entry: uint64 = uint64(parseUInt(line[4 .. indexOfEndBracket - 1]))
        var value: uint64 = uint64(parseUInt(line[indexOfEndBracket + 4 .. line.len - 1]))
        # echo entry, "<-", value

        var correctedEntry = entry
        var floatings = newSeq[uint8]()

        # 36 is the length of the mask
        var thisMask = ' ';
        for i in 0 .. 35:
            thisMask = mask[35 - i]
            if thisMask == '0':
                continue
            elif thisMask == '1':
                correctedEntry.setMask(rotateLeftBits(uint64(1), i))
            else:
                floatings.add(uint8(i))

        # echo floatings

        var correctedEntries = @[correctedEntry]
        var tmpCorrectedEntries = newSeq[uint64]()
        for floating in floatings:
            for correctedEntry in correctedEntries:
                # marked bit at floating to 0
                tmpCorrectedEntries.add(correctedEntry.masked(allOnes - rotateLeftBits(uint64(1), floating)))
                # marked bit at floating to 1
                tmpCorrectedEntries.add(correctedEntry.setMasked(rotateLeftBits(uint64(1), floating)))
            correctedEntries = tmpCorrectedEntries
            tmpCorrectedEntries = @[]

        # echo correctedEntries
        for correctedEntry in correctedEntries:
            memory[correctedEntry] = value

    var total: uint64 = 0
    for value in memory.values:
        total += value

    return total


var input: string = paramStr(1)
# var input = """
# mask = 000000000000000000000000000000X1001X
# mem[42] = 100
# mask = 00000000000000000000000000000000X0XX
# mem[26] = 1
# """

var t0 = cpuTime()
var output = run(input)

echo "_duration:", (cpuTime() - t0) * 1000
echo output
